test/proxy.t.sol

```diff
@@ -52,93 +50,229 @@ contract ProxyTest is DSTest {
+    }
+
+    function testBuildProxyWithAddress(address randomUser) public {
+        vm.assume(randomUser != address(registry));
+        vm.assume(randomUser != address(this));
+        address payable first = registry.build(randomUser);
+        assertEq(Proxy(first).wards(address(this)), 0);
+        assertEq(Proxy(first).wards(randomUser), 1);
+    }
+
+    function testProxyOwnerCantExecute(address randomUser) public {
+        vm.prank(randomUser);
+        address payable proxyAddr = registry.build();
+        Proxy proxy = Proxy(proxyAddr);
+
+        bytes memory data = abi.encodeWithSignature("inlineAdd(uint256,uint256)", 5,7);
 
-        address payable second = registry.build();
-        assertTrue(first != second);
-        proxy = Proxy(second);
-        assertEq(proxy.accessToken(), 2);
+        // set action as a safe target
+        vm.prank(randomUser);
+        proxy.file("target", address(action));
+
+        // executing action by proxy owner fails until they are added as user
+        vm.prank(randomUser);
+        vm.expectRevert(bytes("TinlakeProxy/user-not-authorized"));
+        proxy.userExecute(address(action), data);
+
+        // executing action by proxy owner succeeds once owner is added as user
+        vm.prank(randomUser);
+        proxy.addUser(randomUser);
+        vm.prank(randomUser);
+        proxy.userExecute(address(action), data);
     }
 
-    function testFailInit() public {
-        address payable proxy = registry.build();
-        Proxy(proxy).init(123);
+    function testAddRemoveUser(address user) public {
+        address payable proxyAddr = registry.build();
+        Proxy proxy = Proxy(proxyAddr);
+        proxy.file("target", address(action));
+        bytes memory data = abi.encodeWithSignature("inlineAdd(uint256,uint256)", 5,7);
+
+        // Add a random user
+        proxy.addUser(user);
+        assertEq(proxy.users(user), 1);
+
+        // test executing target actions with user
+        vm.prank(user);
+        proxy.userExecute(address(action), data);
+
+        // remove user
+        proxy.removeUser(user);
+        assertEq(proxy.users(user), 0);
+        vm.prank(user);
+        vm.expectRevert(bytes("TinlakeProxy/user-not-authorized"));
+        proxy.userExecute(address(action), data);
     }
 
-    function testCreate2() public {
-        address payable proxy = registry.build();
-        uint a = Proxy(proxy).accessToken();
-        assertEq(proxy, registry.proxies(a));
+    function testAddTarget(address _target) public {
+        address payable proxyAddr = registry.build();
+        Proxy proxy = Proxy(proxyAddr);
 
-        // second address
-        address payable proxyB = registry.build();
-        uint b = Proxy(proxyB).accessToken();
-        assertEq(proxyB, registry.proxies(b));
+        proxy.file("target", _target);
+        assertEq(proxy.target(), _target);
     }
 
-    function testExecute() public {
+    function testExecute(address user) public {
```

Comment:
These tests should probably all be called `testUserExecute*` rather than `testExecute*`

```suggestion
    function testUserExecute(address user) public {
```